[cxx-interop] [IRGen] TypeInfo for address-only types.

zoecarver opened the pull request:
The current "ClangRecordTypeInfo" derives from "LoadableTypeInfo" and is only meant for loadable types. While we have not yet run into problems, this may cause issues in the future and as more logic is needed around copying, moving, and destroying C++ objects, this should be fixed.

I'm not aware of any behavior changes so, it's a bit difficult for me to come up with tests. But, that may just be a consequence of my lack of knowledge around this part of the codebase. Please chime in with test ideas if you have any :)

Refs #32291.

zoecarver commented on lib/IRGen/GenStruct.cpp:
```diff
@@ -290,7 +291,46 @@ namespace {
+                             fields, storageType, size, std::move(spareBits),
+                             align, IsPOD, IsNotBitwiseTakable, IsFixedSize),
+          ClangDecl(clangDecl) {
+      (void)ClangDecl;
```
I'm sure we'll need this later but, I also don't want compiler warnings.

gribozavr replied:
SGTM.

zoecarver commented on lib/IRGen/GenStruct.cpp:
```diff
@@ -290,7 +291,46 @@ namespace {
-  
+
+  class ClangRecordAddressOnlyTypeInfo final
+      : public StructTypeInfoBase<ClangRecordAddressOnlyTypeInfo, FixedTypeInfo,
```
Can someone more familiar with this part of IRGen confirm that `FixedTypeInfo` is what should go here?

gribozavr replied:
Should we use `IndirectTypeInfo` as the base type?

zoecarver replied:
You mean use `IndirectTypeInfo` instead of `StructTypeInfoBase`? Even address-only types should be imported as structs, so I think we should keep `StructTypeInfoBase`. `IndirectTypeInfo` takes two template parameters so, I think one of those would still have to be `FixedTypeInfo` or something.

gribozavr commented on lib/IRGen/GenStruct.cpp:
```diff
@@ -290,7 +291,46 @@ namespace {
   };
-  
+
+  class ClangRecordAddressOnlyTypeInfo final
```
Could you move it after `ClangRecordTypeInfo` to keep the order in the enum and here identical?

gribozavr commented on lib/IRGen/GenStruct.cpp:
```diff
@@ -290,7 +291,46 @@ namespace {
+    void initializeFromParams(IRGenFunction &IGF, Explosion &params,
+                              Address addr, SILType T,
+                              bool isOutlined) const override {
+      llvm_unreachable("Cannot initialize address-only type from params.");
```
I think we can initialize address-only type from an explosion; the reason why nobody should call this specific method is to avoid violating C++ invariants about an object having a stable address (C++ types that are non-loadable should be created/copied/moved/destroyed only by C++ special member functions). I'd suggest to reword the error to point to that instead.

zoecarver replied:
Yep, you're right.

gribozavr commented on lib/IRGen/GenStruct.cpp:
```diff
@@ -680,6 +720,12 @@ class ClangRecordLowering {
+    if (SwiftType.getStructOrBoundGenericStruct()
+            ->isCxxNotTriviallyCopyable()) {
+      return ClangRecordAddressOnlyTypeInfo::create(
+          FieldInfos, llvmType, TotalStride, std::move(SpareBits),
```
I don't think we can assume any spare bits in a non-trivial C++ type.

zoecarver replied:
Good point.

zoecarver replied:
I had to add this back to avoid an assertion. I will investigate further.

zoecarver replied:
Nevermind, I fixed it.

gribozavr commented on lib/IRGen/GenStruct.cpp:
```diff
@@ -680,6 +720,12 @@ class ClangRecordLowering {
+    if (SwiftType.getStructOrBoundGenericStruct()
+            ->isCxxNotTriviallyCopyable()) {
+      return ClangRecordAddressOnlyTypeInfo::create(
+          FieldInfos, llvmType, TotalStride, std::move(SpareBits),
```
The `FieldInfos` vector will contain private data members that we're not importing. It would be good to skip private data members in `ClangRecordLowering` (in a future PR).

zoecarver replied:
I can make a follow up PR for this.

gribozavr replied:
SGTM!

gribozavr commented on lib/IRGen/GenStruct.cpp:
```diff
@@ -680,6 +720,12 @@ class ClangRecordLowering {
+    if (SwiftType.getStructOrBoundGenericStruct()
+            ->isCxxNotTriviallyCopyable()) {
+      return ClangRecordAddressOnlyTypeInfo::create(
+          FieldInfos, llvmType, TotalStride, std::move(SpareBits),
```
`ClangRecordLowering::addStructField` assumes that all fields will have `LoadableTypeInfo` (by invoking `cast<LoadableTypeInfo>()`) -- but that's no longer true starting with this PR. Are you getting any assertions in the testsuite when processing types that aggregate other types which we consider non-trivial?

zoecarver replied:
Interesting... you're right. No tests caught this so, I need to add a few. I think we would have hit an assertion if we had the tests from #32291. I changed `LoadableTypeInfo` to `FixedTypeInfo` which seemed to work for everything except `getExplosionSize` which I was able to replace with `SwiftDecl->getStoredProperties().size()` for non-loadable types.

zoecarver replied:
I verified the tests I added triggered an error with the old code (`cast<LoadableTypeInfo>()`).

zoecarver commented on lib/IRGen/GenStruct.cpp:
```diff
@@ -53,6 +53,7 @@ enum class StructTypeInfoKind {
   LoadableStructTypeInfo,
   FixedStructTypeInfo,
   ClangRecordTypeInfo,
```
See #33077.

zoecarver committed:
[cxx-interop] [IRGen] TypeInfo for address-only types.

The current "ClangRecordTypeInfo" derives from "LoadableTypeInfo" and is
only meant for loadable types. While we have not yet run into problems,
this may cause issues in the future and as more logic is needed around
copying, moving, and destroying C++ objects, this needs to be fixed.

gribozavr commented on lib/IRGen/GenStruct.cpp:
```diff
@@ -339,6 +340,50 @@ namespace {
+                                   const clang::RecordDecl *clangDecl)
+        : StructTypeInfoBase(StructTypeInfoKind::AddressOnlyClangRecordTypeInfo,
+                             fields, storageType, size,
+                             // Create an empty SpareBitVector with a width of
```
This comment repeats what the code does -- everyone should be able to figure that out. Instead, could you add a comment that explains the reason for it?

zoecarver replied:
👍

gribozavr commented on lib/IRGen/GenStruct.cpp:
```diff
@@ -823,8 +873,10 @@ class ClangRecordLowering {
-  void addFieldInfo(VarDecl *swiftField, const LoadableTypeInfo &fieldType) {
-    unsigned explosionSize = fieldType.getExplosionSize();
+  void addFieldInfo(VarDecl *swiftField, const FixedTypeInfo &fieldType) {
+    unsigned explosionSize = SwiftDecl->getStoredProperties().size();
```
If I understand it correctly, the concept of explosion does not make sense for address-only types, so it is not a surprise that we don't have an appropriate API to call here.

This information is eventually written into `RecordField::Begin` / `::End`. Note that only `ClangFieldInfo` is setting these fields. Even `StructFieldInfo` is not setting them. Since this data is not correct for native Swift structs, I suspect that users of this data are dead code. Could you take a look? It is quite possible you would be able to delete some/all of this stuff.

If this data is used, I think we should either change `ClangFieldInfo` to be like `StructFieldInfo` and not set the explosion begin/end, at least when we have an address-only type.

zoecarver replied:
Yeah, you're right! With a small amount of refactoring, I was able to get this working without `begin`/`end`.

zoecarver replied:
Turns out after running all the tests there are some things which rely on `begin`/`end`. At first, I was confused by this because, as you said, it doesn't appear that `StructFieldInfo` sets them. But, after further investigation, it appears they are set for `StructFieldInfo`. [Just in a more convoluted way.](https://github.com/apple/swift/blob/master/lib/IRGen/GenRecord.h#L868-L871)

I'll have to add back the old path, too.

gribozavr commented on test/Interop/Cxx/class/Inputs/type-classification.h:
```diff
@@ -155,4 +155,22 @@ struct StructDeletedDestructor {
+      : value(other.value) {}
+};
+
+struct StructWithMemberWithCopyConstructor {
```
"StructWithSubobjectCopyConstructorAndValue"

gribozavr commented on test/Interop/Cxx/class/Inputs/type-classification.h:
```diff
@@ -155,4 +155,22 @@ struct StructDeletedDestructor {
+  StructWithCopyConstructorAndValue member;
+};
+
+struct StructWithCopyConstructorAndMemberWithCopyConstructor {
```
"StructWithCopyConstructorAndSubobjectCopyConstructorAndValue"

gribozavr commented on test/Interop/Cxx/class/address-only.swift:
```diff
@@ -0,0 +1,35 @@
+// RUN: %empty-directory(%t)
```
Please rename it to `type-classification-non-trivial.swift` to be in line with other tests.

gribozavr commented on test/Interop/Cxx/class/address-only.swift:
```diff
@@ -0,0 +1,35 @@
+    member: member
+  )
+  expectEqual(obj.member.value, 42)
+}
```
Could you add the relevant patterns to `type-classification-non-trivial-silgen.swift` & make an IRGen test?

zoecarver replied:
Yep. I'll get to this later this week/weekend.

zoecarver replied:
Done :)

gribozavr reviewed (commented):
Thanks for the fixes!

gribozavr commented on lib/IRGen/GenRecord.h:
```diff
@@ -47,7 +47,8 @@ template <class FieldImpl> class RecordField {
   explicit RecordField(const TypeInfo &elementTI)
-    : Layout(ElementLayout::getIncomplete(elementTI)) {}
+      : Layout(ElementLayout::getIncomplete(elementTI)), Begin(0XFFF),
+        End(0XFFF) {}
```
I think you meant `0xFFFF`.

Maybe it is better to leave `Begin` and `End` uninitialized so that MSan can catch a bogus read? Or you could add an `assert(Begin != 0xFFFF);` in `getProjectionRange()`.

zoecarver replied:
Fair enough. Whatever value's left on the stack will probably be a good enough indication anyway.

gribozavr commented on lib/IRGen/GenStruct.cpp:
```diff
@@ -339,6 +346,50 @@ namespace {
+                                   const clang::RecordDecl *clangDecl)
+        : StructTypeInfoBase(StructTypeInfoKind::AddressOnlyClangRecordTypeInfo,
+                             fields, storageType, size,
+                             // An address only type shouldn't assume any spare
+                             // bits. So, give it an empty vector.
```
```suggestion
                             // We can't assume any spare bits in a C++ type with user-defined special member functions.
```

zoecarver commented:
@swift-ci please smoke test

zoecarver commented on test/Interop/Cxx/class/type-classification-non-trivial-irgen.swift:
```diff
@@ -0,0 +1,94 @@
+
+import TypeClassification
+
+// TODO: C++ objects with destructors should be tested here once we fully
```
This is unfortunate (but also unrelated to this change). I was meaning to submit a fix for this and forgot. Maybe I'll have time this weekend. Otherwise, I'll do it next week.

zoecarver replied:
See #32591.

gribozavr commented on test/Interop/Cxx/class/type-classification-non-trivial-irgen.swift:
```diff
@@ -0,0 +1,94 @@
+// TODO: C++ objects with destructors should be tested here once we fully
+// support them.
+
+// CHECK-LABEL: define {{.*}}i1 @"$s4main37testStructWithCopyConstructorAndValueSbyF"
```
Could you add an explanation of what these tests are verifying, so that future when IRGen patterns change people can update the CHECK lines appropriately, and not accidentally eliminate the part that we're thinking we are testing here.

zoecarver commented on test/Interop/Cxx/class/type-classification-non-trivial-silgen.swift:
```diff
@@ -22,4 +22,126 @@ public func testStructWithSubobjectDestructor() {
+// CHECK: [[I_42:%.*]] = struct $Int32 ([[IL_42]] : $Builtin.Int32)
+// CHECK: [[OBJ_VALUE:%.*]] = struct_element_addr [[OBJ]] : $*StructWithCopyConstructorAndValue, #StructWithCopyConstructorAndValue.value
+// CHECK: store [[I_42]] to [[OBJ_VALUE]] : $*Int32
+// CHECK: [[IL_FALSE:%.*]] = integer_literal $Builtin.Int1, -1
```
Done.

zoecarver commented:
@swift-ci please test

zoecarver commented:
@swift-ci please test

zoecarver commented:
@swift-ci please test

zoecarver commented:
@swift-ci please test Windows Platform

zoecarver commented:
@swift-ci please test

zoecarver commented:
@swift-ci please test

gribozavr commented:
@zoecarver Congrats!

meg-gupta commented:
This is causing a Filecheck mismatch error with Interop/Cxx/class/type-classification-non-trivial-irgen.swift  in https://ci.swift.org/job/oss-swift_tools-R_stdlib-RD_test-simulator/5307/ 
@zoecarver can you take a look ?

zoecarver commented:
@meg-gupta thanks for bringing my attention to the issue. I have a fix, running the tests locally now. I'll make a PR shortly.

zoecarver commented:
See #33474.
